src/ranged.cpp

```diff
@@ -1531,3 +1476,66 @@ void splatter( const std::vector<tripoint> &trajectory, int dam, const Creature
         }
     }
 }
+
+void drop_or_embed_projectile( const dealt_projectile_attack &attack )
+{
+    const auto &proj = attack.proj;
+    const auto &drop_item = proj.drop;
+    const auto &effects = proj.proj_effects;
+    if( drop_item == nullptr ) {
+        return;
+    }
+
+    const tripoint &pt = attack.end_point;
+
+    if( effects.count( "SHATTER_SELF" ) ) {
+        // Drop the contents, not the thrown item
+        if( g->u.sees( pt ) ) {
+            add_msg( _("The %s shatters!"), drop_item->tname().c_str() );
+        }
+
+        for( const item &i : drop_item->contents ) {
+            g->m.add_item_or_charges( pt, i );
+        }
+        // TODO: Non-glass breaking
+        // TODO: Wine glass breaking vs. entire sheet of glass breaking
+        sounds::sound(pt, 16, _("glass breaking!"));
+        return;
+    }
+
+    item dropped_item = *drop_item;
+
+    monster *mon = dynamic_cast<monster *>( attack.hit_critter );
+    // Try to embed the projectile in monster
+    // Don't embed on miss, in player/NPCs, when we didn't stab/cut properly
+    //  or when the item simply shouldn't be embedded (for example, it is active)
+    if( mon == nullptr ||
+        ( attack.dealt_dam.type_damage( DT_STAB ) +
+          attack.dealt_dam.type_damage( DT_CUT ) <=
+            attack.dealt_dam.type_damage( DT_BASH ) ) ||
+        effects.count( "NO_EMBED" ) != 0 ) {
+        bool do_drop = true;
+        if( effects.count( "PROCESS_ON_DROP" ) ) {
+            // Don't drop if it exploded
+            do_drop = !dropped_item.process( nullptr, attack.end_point, true );
+        }
+
+        if( do_drop ) {
+            g->m.add_item_or_charges( attack.end_point, dropped_item );
+        }
+
+        if( effects.count( "HEAVY_HIT" ) ) {
+            if( g->m.has_flag( "LIQUID", pt ) ) {
+                sounds::sound( pt, 10, _("splash!") );
+            } else {
+                sounds::sound( pt, 8, _("thud.") );
+            }
+            const trap &tr = g->m.tr_at(pt);
+            if( tr.triggered_by_item( *drop_item ) ) {
+                tr.trigger( pt, nullptr );
+            }
+        }
+    } else {
+        mon->add_item( dropped_item );
```

Comment:
This requires a message. It's very confusing for experienced players that expect the item to appear under the monster.

The PR works technically fine (no idea about the numbers and balancing).